src/ILCompiler.Compiler/src/Compiler/ExportedMethodsRootProvider.cs

```diff
@@ -36,11 +46,44 @@ public void AddCompilationRoots(IRootingServiceProvider rootProvider)
                     if (ecmaMethod.IsNativeCallable)
                     {
                         string nativeCallableExportName = ecmaMethod.GetNativeCallableExportName();
+
                         if (nativeCallableExportName != null)
+                        {
+                            _methods.Add(ecmaMethod);
                             rootProvider.AddCompilationRoot(method, "Native callable", nativeCallableExportName);
+                        }
                     }
                 }
             }
+
+            // _exportsFile is not null when it's a shared library build
+            if (_exportsFile != null)
+                EmitExportedMethods();
+        }
+
+        private void EmitExportedMethods()
+        {
+            string moduleName = Path.GetFileNameWithoutExtension(_exportsFile);
+            FileStream fileStream = new FileStream(_exportsFile, FileMode.OpenOrCreate);
+            using (StreamWriter streamWriter = new StreamWriter(fileStream))
+            {
+                if (_module.Context.Target.IsWindows)
+                {
+                    streamWriter.Write("LIBRARY   ");
```

Comment:
It the `LIBRARY` line actually needed? It seem unnecessary.

Reply:
I created it as specified in the DEF file documentation. I'll do a test and get back to you

Reply:
Number of the .def files in CoreCLR do not have the LIBRARY line. e.g. https://github.com/dotnet/coreclr/blob/master/src/dlls/dbgshim/dbgshim.ntdef

Reply:
Ah, that fine then